Configure/Makefile: fix some FIPS installation issues

mspncp opened the pull request:
This pull request fixes some FIPS installation issues (notably #13679 and #13680), but it is marked as a DRAFT, because

- I need some help from @levitte for the analoge changes to the VMS template (descrip.mms.tmpl) and
- there are still some open questions, ~which need to be written down and resolved first~ see followup post.

For more details, see the commit messages.

**Commits (bottom up)**

Author: Dr. Matthias St. Pierre <matthias.st.pierre@ncp-e.com>
Date:   Tue Dec 15 22:44:32 2020 +0100

    Configure/Makefile: correct the FIPS module configuration file path
    
    According to the OpenSSL 3.0 Wiki, the file should be located at
    
        $(OPENSSLDIR)/fipsmodule.cnf
    
    next to the openssl.cnf file.

Author: Dr. Matthias St. Pierre <matthias.st.pierre@ncp-e.com>
Date:   Tue Dec 15 22:34:41 2020 +0100

    Configure/Makefile: use the correct openssl app for FIPS installation
    
    The `openssl` app was previously called without a path, which
    would generally invoke the system's copy of the openssl application.
    Currently, that's most likely an openssl version 1.1.1 application,
    which does not recognize the `fipsinstall` command and terminates
    with an error message.

Author: Dr. Matthias St. Pierre <matthias.st.pierre@ncp-e.com>
Date:   Tue Dec 15 19:41:58 2020 +0100

    Configure/Makefile: fix the `-macopt` argument of the fipsinstall command
    
    The FIPS hmac key is provided as a hexadezimal string, which needs to
    be be prefixed with `hexkey:`, not `key:`.

Author: Dr. Matthias St. Pierre <matthias.st.pierre@ncp-e.com>
Date:   Tue Dec 15 19:41:37 2020 +0100

    Configure/Makefile fix target dependency for the FIPS provider installation
    
    Don't uninstall entire software just because FIPS provider module configuration
    file is removed.
    
    In more technical terms: since the `install_fips` target depends on `install_sw`,
    the `uninstall_sw` target should depend on `uninstall_fips`, not vice versa.

Author: Dr. Matthias St. Pierre <matthias.st.pierre@ncp-e.com>
Date:   Tue Dec 15 19:27:41 2020 +0100

    Configure/Makefile: fix Perl error in FIPSMODULENAME declaration
    
    Fixes #13679

mspncp commented:
## Open Questions

### FIPS _Module_ or FIPS _Provider_?

The OpenSSL 3.0 Wiki still talks about the [Installation of the FIPS module](https://wiki.openssl.org/index.php/OpenSSL_3.0#Completing_the_installation_of_the_FIPS_Module) and names the configuration file `fipsmodule.cnf`. Shouldn't we call it the FIPS _provider_ and rename the configuration file to `fipsprovider.cnf`, accordingly?

### Life cycle of the FIPS provider

Currently, the FIPS provider is installed unconditionally by the `install_sw` target together with the other providers. The `install_fips` target only creates and installs the configuration file.  If the installation happens to be a service update installation, then the `openssl install` command will invalidate a previously installed FIPS provider.

IIRC, it was the plan to support service updates (bug fixes) of the 'regular' OpenSSL software independently from the FIPS provider. The current installation procedure does not seem to be fit to support that goal. Are we going to fix this and provide a concept how to deal the life cycle and versioning of providers, or are we going to leave the nasty details to the distributors?

One possible step in the right direction would be to exclude the FIPS provider from the `install_sw` target and use the `install_fips` target to for both the provider and the configuration file.

richsalz commented:
@levitte explained this in a PR that I can't find right now.  One thing to consider is that the "FIPS Module" is openssl's, and it is a FIPS provider.  There can be other modules from third parties, etc.

levitte commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -561,15 +562,14 @@ uninstall_docs: uninstall_man_docs uninstall_html_docs
-	@openssl fipsinstall -module $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME) \
-		-out $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf \
-		-macopt 'key:$(FIPSKEY)'
+	$(SRCDIR)/util/wrap.pl $(SRCDIR)/apps/openssl fipsinstall \
```
`$(SRCDIR)/util/wrap.pl $(SRCDIR)/apps/openssl` should be `$(BLDDIR)/util/wrap.pl $(BLDDIR)/apps/openssl`

mspncp replied:
Yeah, right. I forgot about the out-of-tree builds. Since the hunks of the last two commits collide, I integrated and cleaned up the fixups immediately and [force-pushed](https://github.com/openssl/openssl/compare/d87bc00811becea20e85015b91ed3faea158f810..e45552c442351cc6eec3b4515e1890a2c5378505) the result.

levitte commented on Configurations/windows-makefile.tmpl:
```diff
@@ -464,23 +464,22 @@ depend:
-	@openssl fipsinstall -module $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME) \
-		-out $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf \
-		-macopt "key:$(FIPSKEY)"
+	"$(SRCDIR)\util\wrap.pl" "$(SRCDIR)\apps\openssl" fipsinstall \
```
`"$(SRCDIR)\util\wrap.pl" "$(SRCDIR)\apps\openssl"` should be `"$(BLDDIR)\apps/openssl"`

(uitl/wrap.pl is Unix only)

mspncp replied:
I thought you had intended the Perl Script to be a platform independent wrapper, when you created it (contrary to shlib_wrap.sh). Even on windows, it's not a no-op, it sets up the module paths. So thought I might as well leave it for symmetry and future extensibility. But if you think it's pointless here, I can remove it.

https://github.com/openssl/openssl/blob/0cc0164d193f6071a9d06b2116a410f8139a7e3c/util/wrap.pl#L9-L12

Although I have to admit I _do_ like the way it treats the Windows case:

https://github.com/openssl/openssl/blob/0cc0164d193f6071a9d06b2116a410f8139a7e3c/util/wrap.pl#L22-L27

levitte replied:
Ah, duh, you're right.  I had just forgotten...

levitte commented:
> ## FIPS _Module_ or FIPS _Provider_?

"FIPS Provider Module"

*Module* is a dynamically loadable thing (so an ENGINE is a *module* as well).

In that sense, the default provider isn't a module, 'cause it's built into libcrypto, while the legacy provider normally is a module.

The FIPS provider will always be a module, so we're not terribly picky when we talk about it, "module" is implied.

levitte commented:
> IIRC, it was the plan to support service updates (bug fixes) of the 'regular' OpenSSL software independently from the FIPS provider. The current installation procedure does not seem to be fit to support that goal. Are we going to fix this and provide a concept how to deal the life cycle and versioning of providers, or are we going to leave the nasty details to the distributors?
> 
> One possible step in the right direction would be to exclude the FIPS provider from the `install_sw` target and use the `install_fips` target to for both the provider and the configuration file.

Good questions.  It does make sense to install it separately, but [sigh] another special case.  Oh well.

mspncp commented:
@levitte would you mind providing the VMS equivalents to [this change](https://github.com/openssl/openssl/pull/13684/commits/c03c9e9272a5a4a53816201d467c9040a8976f84) and [that change](https://github.com/openssl/openssl/pull/13684/commits/e45552c442351cc6eec3b4515e1890a2c5378505), so I can amend the respective commits? Thanks in advance!

mspncp commented:
Since the two questions in https://github.com/openssl/openssl/pull/13684#issuecomment-745609203 are now answered respectively moved into a new issue #13693, this pull request will be ready for review as soon as I get the VMS ports from @levitte.

rsbeckerca commented:
Is there anything needed for the NonStop port for this PR?

levitte commented:
> Is there anything needed for the NonStop port for this PR?

I believe we have it covered, no?

rsbeckerca commented:
> > Is there anything needed for the NonStop port for this PR?
> 
> I believe we have it covered, no?

Hopefully. Waiting on the next alpha/beta.

mspncp commented:
> > > Is there anything needed for the NonStop port for this PR?

> Hopefully. Waiting on the next alpha/beta.

Assuming that your Makefile on the NonStop platform is generated from the `Configurations/unix-Makefile.tmpl` template during the configure step (as it is the case for unix-like platforms), you should be fine.

mspncp commented:
During rebase, I dropped commit ab82edb924 (Configure/Makefile: fix target dependency for the FIPS provider installation) and therefore added 9961684dbc3fac3cec3fd3c693a7645d4496eb59, which is co-authored by @levitte and me: I cherry-picked his changes from #13696 and added the corresponding changes for the Windows Makefile.

The corresponding changes for VMS (Configurations/descrip.mms.tmpl) are still missing. Nevertheless, I marked it ready-for review. @levitte maybe you want to add the VMS changes yourself in a followup pull request?

mspncp commented:
Currently testing on Windows...

levitte commented:
Re VMS, there's more stuff that needs to happen (and I have a branch), so I suggest leaving it out of this change and I'll get back to it as part of the other updates I've in the works.

mspncp commented:
@levitte I switched from ActiveState Perl to Strawberry Perl (as recommended by the docs), but I'm still getting this annoying error about the missing "basename" method in the Makefile after configuring. What am I missing?

```
FIPSMODULENAME=Program fragment delivered error ``Can't locate object method "basename" via package "platform" at Configurations\windows-makefile.tmpl line 60.''
```
Here's my perl version and my [configdata dump](https://github.com/openssl/openssl/files/5869674/configdata.txt).

```
C:\src\openssl>where perl
C:\Strawberry\perl\bin\perl.exe

C:\src\openssl>perl --version

This is perl 5, version 32, subversion 0 (v5.32.0) built for MSWin32-x64-multi-thread

Copyright 1987-2020, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.
` `

levitte reviewed (dismissed):
Ooooops

mspncp commented:
> @levitte I switched from ActiveState Perl to Strawberry Perl (as recommended by the docs), but I'm still getting this annoying error about the missing "basename" method in the Makefile after configuring. What am I missing?
> 
> ```
> FIPSMODULENAME=Program fragment delivered error ``Can't locate object method "basename" via package "platform" at > Configurations\windows-makefile.tmpl line 60.''
> ```

This Perl error is blocking the PR an I have no idea why I am missing the `basename` method from the `platform` package. It's a vanilla Strawberry Perl installation on Windows and I installed `Text::Template` and `Test::More` using CPAN, as documented in the  Perl notes. Is there something I missed in the installation instructions?

mspncp commented:
Ok @levitte, I found the problem: it turned out that your suggestion https://github.com/openssl/openssl/issues/13679#issuecomment-744774350 was correct, but my fix in #13684
 (commit a8817257637c5935dff51ea08cbec70c1d60cade) actually fixed *all* of the three makefiles, *except for* the Windows makefile. :-/

I pushed a fixup to #13684, (commit 9c9741f16fd43222081085cc4f1653527d5c43aa).

mspncp commented:
Something is still wrong @levitte: building and installing...

```
perl Configure VC-WIN64A shared --prefix=\dev\openssl --openssldir=\dev\openssl\ssl
nmake
nmake install
```
succeeds (see [build.bat](https://github.com/openssl/openssl/files/5922249/build.bat.txt)  and [build.log](https://github.com/openssl/openssl/files/5922250/build.log)), but afterwards `nmake install_fips` fails with the following error:

```
C:\src\openssl>nmake install_fips

Microsoft (R) Program Maintenance Utility, Version 14.28.29336.0
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.

*** Installing FIPS module
install engines\capi.dll engines\padlock.dll -> \dev\openssl\lib\ossl-modules\fips.dll
Das System kann den angegebenen Pfad nicht finden.
NMAKE : fatal error U1077: "copy": RÃ¼ckgabe-Code "0x1"
Stop.
```

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

mspncp commented:
Last force-push was a plain rebase on master in order to pick up #14067, except for the [last commit](https://github.com/openssl/openssl/pull/13684/commits/29f21557e3283dbe422d8aadf3a58a1899472751), which fixes the aforementioned error:

> `nmake install_fips` fails with the following error:
> 
> ```
> C:\src\openssl>nmake install_fips
> 
> Microsoft (R) Program Maintenance Utility, Version 14.28.29336.0
> Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.
> 
> *** Installing FIPS module
> install engines\capi.dll engines\padlock.dll -> \dev\openssl\lib\ossl-modules\fips.dll
> Das System kann den angegebenen Pfad nicht finden.  #  (= ERROR_PATH_NOT_FOUND, mspncp)
> NMAKE : fatal error U1077: "copy": RÃ¼ckgabe-Code "0x1"
> Stop.
> ```

mspncp commented on Configurations/windows-makefile.tmpl:
```diff
@@ -105,7 +105,7 @@ INSTALL_ENGINEPDBS={-
 INSTALL_FIPS={-
         join(" ", map { platform->dso($_) }
```
@levitte INSTALL_ENGINES and other INSTALL_XXX variables quote their arguments ( `quotify1(platform->dso($_))`), while INSTALL_FIPS and INSTALL_MODULES don't do it. Is that on purpose, or an oversight?

mspncp replied:
Looking at the generated Makefile I wonder whether it wouldn't be better to remove all those unnecessary quotes instead. They don't seem to contain user configurable parts which could contain spaces.

```
INSTALL_LIBS="libcrypto.lib" "libssl.lib"
INSTALL_SHLIBS="libcrypto-3-x64.dll" "libssl-3-x64.dll"
INSTALL_SHLIBPDBS="libcrypto-3-x64.pdb" "libssl-3-x64.pdb"
INSTALL_ENGINES="engines\\capi.dll" "engines\\padlock.dll"
INSTALL_ENGINEPDBS="engines\\capi.pdb" "engines\\padlock.pdb"
INSTALL_FIPS=providers\fips.dll
INSTALL_MODULES=providers\legacy.dll
INSTALL_MODULEPDBS="providers\\fips.pdb" "providers\\legacy.pdb"
INSTALL_PROGRAMS="apps\\openssl.exe"
INSTALL_PROGRAMPDBS="apps\\openssl.pdb"
BIN_SCRIPTS="tools\\c_rehash.pl"
MISC_SCRIPTS="apps\\CA.pl" "apps\\tsget.pl"
HTMLDOCS1=doc\html\man1\CA.pl.html doc\html\man1\openssl-asn1parse.html 
```

levitte replied:
It's an oversight...  INSTALL_FIPS and INSTALL_MODULES should have their arguments quote too.

> They don't seem to contain user configurable parts which could contain spaces

Believe you me, users may surprise us still...

mspncp replied:
Done in d76ccbdf1b6d04591fff83b28a2233dd198c4e89.

mspncp commented:
I removed the `approval: done` because I'm still hoping that @levitte will take a look at it some time.

mspncp commented:
The last commit was cherry-picked from #13684 and will be dropped from this pull request before merging.

rsbeckerca commented:
Everything is currently looking fine for the NonStop build. I'll look for the merge and retest ASAP after that.

t8m reviewed (commented):
This needs some documentation change in INSTALL.md as the install_fips is mentioned there with different semantics. Also what about the VMS template? Should  you just open an issue after merging this?

t8m commented:
Other than the documentation and VMS, this really LGTM.

levitte commented:
> Also what about the VMS template?

It needs separate work anyway.  And yeah, I should raise an issue.

rsbeckerca commented:
Is there anything needed for NonStop (x86) or are we fine if standard tests continue to pass?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

mspncp commented:
Rebased and cleaned up. No sure how much of this pr is still relevant because I haven't been able to pay much attention to the recent development.

t8m commented:
The external test failure is irrelevant but the Windows minimal build failure is IMO relevant.

t8m commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -587,17 +596,27 @@ install_docs: install_man_docs install_html_docs
+	@$(ECHO) "fipsinstall $(OPENSSLDIR)/fipsmodule.cnf"
+	@$(BLDDIR)/util/wrap.pl $(BLDDIR)/apps/openssl fipsinstall \
+		-module $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME) \
+		-out $(OPENSSLDIR)/fipsmodule.cnf \
```
Should this be rather `$(DESTDIR)$(OPENSSLDIR)/fipsmodule.cnf`

mspncp replied:
Thanks, you are right.

mspncp replied:
Fixed in def19f6dd37c7e4c973c664926665129fc673af5 (all locations).

t8m commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -587,17 +596,27 @@ install_docs: install_man_docs install_html_docs
-	@openssl fipsinstall -module $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME) \
-		-out $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf \
-		-macopt 'key:$(FIPSKEY)'
+	@$(ECHO) "fipsinstall $(OPENSSLDIR)/fipsmodule.cnf"
```
`$(DESTDIR)$(OPENSSLDIR)/fipsmodule.cnf`

t8m commented on Configurations/unix-Makefile.tmpl:
```diff
@@ -587,17 +596,27 @@ install_docs: install_man_docs install_html_docs
 	@$(ECHO) "*** Uninstalling FIPS module configuration"
-	@$(ECHO) "$(RM) $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf"
-	@$(RM) $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf
+	$(RM) $(OPENSSLDIR)/fipsmodule.cnf
```
`$(DESTDIR)$(OPENSSLDIR)/fipsmodule.cnf`

t8m commented on Configurations/windows-makefile.tmpl:
```diff
@@ -471,17 +478,26 @@ install_docs: install_html_docs
-	@openssl fipsinstall -module $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME) \
-		-out $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf \
-		-macopt "key:$(FIPSKEY)"
+	@$(ECHO) "fipsinstall $(OPENSSLDIR)\fipsmodule.cnf"
```
Add `$(DESTDIR)` ?

t8m commented on Configurations/windows-makefile.tmpl:
```diff
@@ -471,17 +478,26 @@ install_docs: install_html_docs
+	@$(ECHO) "fipsinstall $(OPENSSLDIR)\fipsmodule.cnf"
+	@$(BLDDIR)\util\wrap.pl $(BLDDIR)\apps\openssl fipsinstall \
+		-module "$(DESTDIR)$(MODULESDIR)\$(FIPSMODULENAME)" \
+		-out "$(OPENSSLDIR)\fipsmodule.cnf" \
```
Add `$(DESTDIR)` ?

mspncp replied:
See 622f240be4d8b43afc16670d7dd15e4356eb3327.

t8m replied:
Ah, right.

t8m commented on Configurations/windows-makefile.tmpl:
```diff
@@ -471,17 +478,26 @@ install_docs: install_html_docs
 	@$(ECHO) "*** Uninstalling FIPS module configuration"
-	@$(ECHO) "$(RM) $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf"
-	@$(RM) $(DESTDIR)$(MODULESDIR)/$(FIPSMODULENAME).cnf
+	$(RM)  "$(OPENSSLDIR)\fipsmodule.cnf"
```
Add `$(DESTDIR)` ?

t8m reviewed (commented):
@levitte perhaps should look as well

rsbeckerca commented:
Would not every test referencing fipsmodule.cnf also have to change? This looks fairly high impact to me. I did think we got this working in a separate PR already merged.

mspncp commented:
> Would not every test referencing fipsmodule.cnf also have to change? This looks fairly high impact to me. I did think we got this working in a separate PR already merged.

If you could figure out which PR this was, that we great. As I said, I was not able to follow the work on master for several weeks because I was too occupied with my regular work.

rsbeckerca commented:
> 
> 
> > Would not every test referencing fipsmodule.cnf also have to change? This looks fairly high impact to me. I did think we got this working in a separate PR already merged.
> 
> If you could figure out which PR this was, that we great. As I said, I was not able to follow the work on master for several weeks because I was too occupied with my regular work.

I think it was #14623

mspncp commented:
@rsbeckerca I don't think this change interferes with testing at all. It only changes the command which gets executed by the `make install_fips` target.

rsbeckerca commented:
> 
> 
> @rsbeckerca I don't think this change interferes with testing at all. It only changes the command which gets executed by the `make install_fips` target.

It was the command itself that was broken before. I had to put $(PERL) in, and @levitte changed how the output file was written. I'm planning on retesting the whole thing (6 builds across 2 system types) once its merged.

richsalz commented:
> and @levitte (I think) changed how the output file was written.

Me; #14623

rsbeckerca commented:
> 
> 
> > and @levitte (I think) changed how the output file was written.
> 
> Me; #14623

Thanks. I'm having a rough technology day :) That and I shouldn't reply from my phone.

mspncp commented:
> > and @levitte (I think) changed how the output file was written.
> 
> Me; #14623

So do you see any interference between this PR and your #14623 @richsalz?

mspncp commented:
> I assume that the build failure of the Windows minimal build is the same as the one in #14136, since I cherry-picked the commit.

I fixed the problem in #14136 and cherry-picked it temporarily as  81973b9f9da626a81185be5d44add42f9d0b4ca5.

t8m commented on README-FIPS.md:
```diff
@@ -28,12 +28,21 @@ The FIPS module must have the self tests run, and the FIPS module config file
-and module path accordingly.
+If you configured OpenSSL to be installed to a different location, the paths will
+vary accordingly. In the rare case that you need to install the fipsmodule.cnf
+to a differnt location, you can execute the `openssl fipsinstall` command manually.
```
different

t8m reviewed (approved):
LGTM now.

richsalz commented:
Using `-out FOO` versus no flag and `> FOO` is a matter of style.  As long as it's not `-out -` then I don't expect any problems. You are dropping the explicit `$(PERL)` before the "wrap" which @rsbecker might be upset about.

rsbeckerca commented:
> Using `-out FOO` versus no flag and `> FOO` is a matter of style.  As long as it's not `-out -` then I don't expect any problems. You are dropping the explicit `$(PERL)` before the "wrap" which @rsbecker might be upset about.

Please do not remove the $(PERL). That and the -out thing cost 2 weeks of debugging to track down the breakage.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

mspncp commented:
> So please @mspncp use $(PERL) here.

No problem, will do. Also, I'm going to cleanup the history and properly rebase this PR onto #14136 in order to get rid of the two cherry-picks from that PR.

mspncp commented:
In the last [force-push](https://github.com/openssl/openssl/compare/81973b9f9da626a81185be5d44add42f9d0b4ca5..db420fcf773cf616b97b2a9c56a7d6fd03f82d63), I dropped the two cherry-picks and rebased onto #14136 instead. The tree changes are caused by the fact that the two cherry-picks were outdated. All commits from this PR are unchanged. To be continued...

rsbeckerca commented:
As a baseline for this, build/test is fine on NonStop x86 and ia64 as of commit 28fd895305.

mspncp commented:
@t8m cleaning up the history turned out to be rather tricky, because the fixups didn't commute nicely. In fact, the hunks of the commits are so close, that every edit of a commit (via interctive rebase) almost always resulted in merge conflicts of the following commits. After two failed attempts, I was close to give up and squash everything together.

But finally, I managed to get it right (hopefully). Please take a look at the [force-push](https://github.com/openssl/openssl/compare/db420fcf773cf616b97b2a9c56a7d6fd03f82d63..d80538ebf42ab52a4ed2d75d633dd04c8f531dfb) diff for the changes since your last review, and double check the cleaned up commit history.

mspncp commented:
> As a baseline for this, build/test passes on NonStop x86 and ia64 as of commit [28fd895](https://github.com/openssl/openssl/commit/28fd8953059fe7d9acd57ef6620457cb41a80509).

Please note that the merge-base of this PR is still on commit 9695f6de1579f5d46e75cfebbaf44bc99cb421ec, which is behind the tip of master by 48 commits. I kept the merge-base on purpose to keep reviewing simple for @t8m.

rsbeckerca commented:
> > As a baseline for this, build/test passes on NonStop x86 and ia64 as of commit [28fd895](https://github.com/openssl/openssl/commit/28fd8953059fe7d9acd57ef6620457cb41a80509).
> 
> Please note that the merge-base of this PR is still on commit 9695f6de1579f5d46e75cfebbaf44bc99cb421ec, which is behind the tip of master by 48 commits. I kept the merge-base on purpose to keep reviewing simple for @t8m.

I think it is prudent to wait for the merge and retest, then fix if needed. I'm in a brutal Sprint right now anyway. Will have debug cycles available by the weekend.

mspncp commented:
> I'm in a brutal Sprint right now anyway.

Well good look then for your 'brutal' sprint.

![image](https://user-images.githubusercontent.com/11467854/114473756-17970c80-9bf5-11eb-8479-5fd8705eebb8.png)

t8m reviewed (dismissed):
Still OK

mspncp commented:
Last force-push was a pure rebase onto the HEAD of master, to get rid of the commit from #141136. No re-review required IMHO.

mspncp commented:
@rsbeckerca since this PR has been rebased now and is waiting for the grace period to pass, would you mind testing it on NonStop bevore it gets merged?

rsbeckerca commented:
Will do

rsbeckerca commented:
> 
> 
> @rsbeckerca since this PR has been rebased now and is waiting for the grace period to pass, would you mind testing it on NonStop bevore it gets merged?

Build/test is in progress.

rsbeckerca commented:
> 
> 
> > @rsbeckerca since this PR has been rebased now and is waiting for the grace period to pass, would you mind testing it on NonStop bevore it gets merged?
> 
> Build/test is in progress.

Passed!

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

mspncp commented:
There is still an issue which I noted while testing on Windows, so I'm not going to merge yet.

t8m commented:
> There is still an issue which I noticed while testing on Windows, so I'm not going to merge yet.

And that issue is not there without this PR?

mspncp commented:
No, it's not a new issue, it's the fact that this fix turned out to be incomplete:

First of all, I noticed that removing _all_ dependencies from the `install_fips` target was overdressed, it has the negative side effect that you can't invoke `make install_fips` directly after configuring. When looking for a (minimal) dependency, I noticed that there is another fips configuration target in the makefile (on which the tests depend):

```make
tests: providers\fipsmodule.cnf
providers\fipsmodule.cnf: "providers\$(PERL)" "providers\fips"
	"$(PERL)" providers\$(PERL) ../util/wrap.pl ../apps/openssl fipsinstall -module providers/$(FIPSMODULENAME) -provider_name fips -mac_name HMAC -section_name fips_sect > $@
```	

This fipsinstall call doesn't stem directly from the Windows makefile template, it is generated from the provider build info.

https://github.com/openssl/openssl/blob/6cf2e8535f64090e822b589fc3402382bea9e1de/providers/build.info#L151-L153

That call uses the openssl app from the source tree, but doesn't wrap the call, so it will fail if the preinstalled openssl.cnf is faulty (which was the starting point of this pull request).

Adding the wrapper script to `providers/build.info` would be easy, but adding the `$(PERL)` in front of it is not possible, because the configure script considers the first item of the generate action to be the generator file, which it treats specially. So from

```
  GENERATE[fipsmodule.cnf]=$(PERL) ../util/wrap.pl ../apps/openssl fipsinstall \
        -module providers/$(FIPSMODULENAME) -provider_name fips \
        -mac_name HMAC -section_name fips_sect
```

the following lines are created:

```make
providers\fipsmodule.cnf: "providers\$(PERL)" "providers\fips"
	"$(PERL)" providers\$(PERL) ../util/wrap.pl ../apps/openssl fipsinstall -module providers/$(FIPSMODULENAME) -provider_name fips -mac_name HMAC -section_name fips_sect > $@
```	


Two thoughts come to my mind:

- If I add the wrapper script without the `$(PERL)`, @rsbeckerca's tests on NoneStop might fail again. But I don't know how to add it in the build.info.
- If there is a `providers\fipsmodule.cnf` which gets created during the regular build, then the `make install_fips` target should simply depend on it and copy it to the final location instead of recreating another one, right?

mspncp commented:
_(note that my last reply was faulty, I had to edit it)_

t8m commented:
@mspncp If you look at how the providers/fipsmodule.cnf target is generated in the current Makefile without any changes, you'll see something like this:
```
providers/fipsmodule.cnf: apps/openssl providers/fips.so $(BLDDIR)/util/wrap.pl
        $(PERL) $(BLDDIR)/util/wrap.pl apps/openssl fipsinstall -module providers/$(FIPSMODULENAME) -provider_name fips -mac_name HMAC -section_name fips_sect > $@
```

So no worry, this is already properly handled by the Makefile generation Perl magic :grin:.

mspncp commented:
Good to know that at least it's intendended to work. But are you sure that it does work on Windows as well? Here is the rule generated from the unmodified `build.info`:

```make
providers\fipsmodule.cnf: apps\openssl.exe providers\fips.dll
	apps\openssl.exe fipsinstall -module providers/$(FIPSMODULENAME) -provider_name fips -mac_name HMAC -section_name fips_sect > $@
```
Consequently, I did get the broken-preinstalled-openssl.cnf error when running `make tests`.

Maybe @levitte added the wrapper only for posix platforms? I recall reminding him only recently, that his `wrap.pl` was intended to be used for all platforms, in particular because it sets module paths and now also the configuration file path.

t8m commented:
Ah, yeah, the wrap.pl is called in https://github.com/openssl/openssl/blob/master/Configurations/unix-Makefile.tmpl#L1452
but missing in https://github.com/openssl/openssl/blob/master/Configurations/windows-makefile.tmpl#L766 and https://github.com/openssl/openssl/blob/master/Configurations/descrip.mms.tmpl#L953

rsbeckerca commented:
@mspncp let me know when you want a retest, please. I'm tracking your branch.

mspncp commented:
> Ah, yeah, the wrap.pl is called in https://github.com/openssl/openssl/blob/master/Configurations/unix-Makefile.tmpl#L1452
> but missing in https://github.com/openssl/openssl/blob/master/Configurations/windows-makefile.tmpl#L766 and https://github.com/openssl/openssl/blob/master/Configurations/descrip.mms.tmpl#L953

You are looking in the wrong place, the wrapper not missing in `windows-makefile.tmpl` anymore.
https://github.com/openssl/openssl/blob/6cf2e8535f64090e822b589fc3402382bea9e1de/Configurations/windows-makefile.tmpl#L491-L492

It is the perl magic code which deals with the build.info entry into the make rule which adds the wrapper for Linux but omits it for Windows.

https://github.com/openssl/openssl/blob/6cf2e8535f64090e822b589fc3402382bea9e1de/providers/build.info#L151-L153


_(VMS is untouched, @levitte will fix it at some time hopefully.) _

mspncp commented:
Sorry @t8m for replying before following your links. Yes, you are completely right, thanks for pointing me to the correct location. I'll fix it.

mspncp commented:
Ooohhh... @rsbeckerca finally I understand what you were talking about in https://github.com/openssl/openssl/pull/13684#issuecomment-816171430 and why you were was so anxious about the `$(PERL)` and the test-related fips configuration in https://github.com/openssl/openssl/pull/13684#issuecomment-816670507! 

You effectively fixed my problem on Linux by adding the Perl wrapper in commit 92a979b4034cd6c1da9cc71736929eb4161359d2. If you had pointed me to that commit then it probably wouldn't have taken me so long to understand.

mspncp commented:
@t8m hopefully this is the last iteration. Essentially I added two commits

Date:   Wed Apr 14 20:23:43 2021 +0200

    build.info: add the Perl wrapper to build generator programs on Windows
    
    Pull request #14320 introduced the ability to use compiled programs
    as generators in GENERATE rules of build.info files. Those generator
    calls were wrapped by the Perl wrapper (wrap.pl) in the Unix makefile
    template, but not on Windows.
    
    This commit adds the missing wrapper for Windows, because for the
    `fipsmodule.cnf` target it is essential that the `openssl fipsinstall`
    command does not load any preinstalled openssl configuration file.


Date:   Wed Apr 14 20:37:37 2021 +0200

    Configure/Makefile: don't generate a fresh fipsmodule.cnf when installing it
    
    There is already a `providers/fipsmodule.cnf` target which is required by
    the tests. Instead of creating another fipsmodule.cnf, the `install_fips`
    target simply copies that configuration file to its final destination.
    
    This commit also restores the minimal dependencies to build the `install_fips`
    target immediately after configuring, which was broken after the removal
    of the `install_sw` dependency.

Those two commits supersede some work of the previous commits, but after my horrible experiences with rewriting the history of this PR, I refrained from doing it again.

mspncp commented:
> @mspncp let me know when you want a retest, please. I'm tracking your branch.

@rsbeckerca now would be a good moment to retest.

rsbeckerca commented:
> 
> 
> > @mspncp let me know when you want a retest, please. I'm tracking your branch.
> 
> @rsbeckerca now would be a good moment to retest.

Kicked off the build/test cycle.

rsbeckerca commented:
> 
> 
> > > @mspncp let me know when you want a retest, please. I'm tracking your branch.
> > 
> > 
> > @rsbeckerca now would be a good moment to retest.
> 
> Kicked off the build/test cycle.

NonStop x86 looks good. ia64 kicked off.

t8m commented on Configurations/windows-makefile.tmpl:
```diff
@@ -762,8 +775,8 @@ EOF
           return <<"EOF";
-$args{src}: $gen0 $deps
-	$gen0$gen_args > \$@
+$args{src}: $gen0 $deps "\$(BLDDIR)/util/wrap.pl"
```
Should backslashes be used here? But they would have to be probably doubled?

mspncp replied:
I thought of using backslashes, but since it's not an ordinary make rule but part of the Perl code I didn't dare to do it because I don't really know how many levels of quoting it would require inside the string (perl, make, cmd, ...). But I can give it another try.

I also noticed that the Windows makefile is terribly inconsistent about pathes need to be quoted and which not (e.g., those containing `$(SRCDIR)` or `$(BLDDIR)`).

https://github.com/openssl/openssl/blob/60caee4bf1c780e95e2853b785e94ef65c8f2621/Configurations/windows-makefile.tmpl#L610-L614

mspncp replied:
Converted to backslashes in 55e52aabd1aa27d1749eb099c0803d61bb49b06c.

rsbeckerca commented:
> 
> 
> > > > @mspncp let me know when you want a retest, please. I'm tracking your branch.
> > > 
> > > 
> > > @rsbeckerca now would be a good moment to retest.
> > 
> > 
> > Kicked off the build/test cycle.
> 
> NonStop x86 looks good. ia64 kicked off.

ia64 passed (FIPS is not a factor on this platform).

paulidale commented:
Any word on this progressing @mspncp?

mspncp commented:
> Any word on this progressing @mspncp?

Current state is: stalled while responding to [Tomas' comment](https://github.com/openssl/openssl/pull/13684#discussion_r613813719) and trying to do some final testing. (Testing on Windows is a PITA due to #14055.) Hopefully I'll find some time in the next few days to finish this up.

mspncp commented:
Ok, Windows and Linux seem to work as far as I can tell. So it's ready for final review.  I'd appreciate some independent testing, in particular whether the issues addressed in #13693 can be considered resolved with this PR.

FIPS support for VMS is out of my skope, @levitte will have to deal with it in a separate PR.

t8m commented:
My concern is that it would be surprising if in kind of "normal" installation you will not get the fips module installed. IMO it should be installed unless you build with no-fips which is what is happening already.

So if you want to use the fips provider from a different openssl build you simply configure and build with no-fips and with that other openssl you use the install_fips instead of the full install.

So IMO the adjustment needed here is to add install_fips to the install_sw target.

mspncp commented:
> So IMO the adjustment needed here is to add install_fips to the install_sw target.

Wasn't it our goal to be able to update an OpenSSL installation without overwriting a validated FIPS provider, for which there is no updated certified version yet?  (#13693, see also see also commit 85ad7f9bc06f65e3f017cdaf2afdf876b56597ca ) With your suggestion, that wouldn't be possible anymore, would it?

t8m commented:
I do not interpret #13693 like that. You are saying we should not overwrite the installed fips provider unconditionally. And I have no problem with that but IMO the condition should exactly be the `no-fips` on the Configure command line. And _maybe_ we should make the `fips` Configure option disabled by default.

In other words if I spend the time to build the fips provider I most probably also want it installed if I am installing the openssl build.

rsbeckerca commented:
> 
> 
> I do not interpret #13693 like that. You are saying we should not overwrite the installed fips provider unconditionally. And I have no problem with that but IMO the condition should exactly be the `no-fips` on the Configure command line. And _maybe_ we should make the `fips` Configure option disabled by default.
> 
> In other words if I spend the time to build the fips provider I most probably also want it installed if I am installing the openssl build.

That's a pretty significant requirements change, IMO. This would for all intents and purposes require having one of `fips` or `no-fips` Configuration option. On the other hand, we have seen (on NonStop) if `fips` is configured but not installed, many tests fail - if this is a property of the tests only, the tests should all be fixed, but if this is behaviour of OpenSSL, then installing `fips` is mandatory.

t8m commented:
> if `fips` is configured but not installed, many tests fail

The test should pass without installing anything, if by `installed` you mean invoking `make install`.
If, on NONSTOP, this does not work, then it is a bug on that platform. But I suppose you actually mean something else by `installed`.

rsbeckerca commented:
> 
> 
> > if `fips` is configured but not installed, many tests fail
> 
> The test should pass without installing anything, if by `installed` you mean invoking `make install`.
> If, on NONSTOP, this does not work, then it is a bug on that platform. But I suppose you actually mean something else by `installed`.

This was discussed in #14612 and caused a cascade failure. There was a platform issue but it was resolved per the PR. However, if the FIPS install fails during `make test`, there is a cascade issue with other tests that assume that the FIPS install did not fail. The `providers/fips_modules.cnf` ended up being empty, which I think caused the cascade. If it wasn't there, perhaps the tests would have worked.

t8m commented:
I think you are talking about something completely unrelated to the final fips module installation step to the final destination. (That is what make install_fips is doing.) You are talking about providers/fipsmodule.cnf creation and that is unrelated.

mspncp commented:
> In other words if I spend the time to build the fips provider I most probably also want it installed if I am installing the openssl build.

@t8m I never thought of it that way and I must say I like your point of view very much: it makes perfect sense that the sequence

```
./config <options> <target> && make && make install
```

builds and installs openssl exactly with the features you have selected. Maybe I was just thinking too complicated. OTOH, for developers it might still be useful to have a low-level `install_fips` target similar to `install_sw` and `install_docs`. In view of what you said, I'm going to reconcider whether the change 85ad7f9bc06f65e3f017cdaf2afdf876b56597ca is useful for developers or not.


> And _maybe_ we should make the `fips` Configure option disabled by default.

I also like your proposal to have the fips provider disabed by default. After all, enabling it has a heavy impact on build and install times and the majority of the countries in the world does not care so much about fips. We should talk about that on our next meeting.

mspncp commented:
@rsbeckerca I agree with @t8m, you and we were talking about different things: You were talking about present or past bugs, we were considering the intended behaviour for the final release.

I understand you are anxious that we don't break anything last-minute by changing some behaviour. However, if tests or the installation fail on a supported platform , just because some specific combination of options is selected, that' simply a bug which needs to be fixed and not an intended behaviour.

@t8m's suggestion to follow the principle: "what you configured is what you get installed" is very natural and more intuitive than having to execute a separate `make install_fips` step manually.

rsbeckerca commented:
> 
> 
> @rsbeckerca I agree with @t8m, you and we were talking about different things: You were talking about present or past bugs, we were considering the intended behaviour for the final release.
> 
> I understand you are anxious that we don't break anything last-minute by changing some behaviour. However, if tests or the installation fail on a supported platform , just because some specific combination of options is selected, that' simply a bug which needs to be fixed and not an intended behaviour.

I'll keep monitoring and fix where needed :)

> @t8m's suggestion to follow the principle: "what you configured is what you get installed" is very natural and more intuitive than having to execute a separate `make install_fips` step manually.

I agree with that.

t8m commented:
Please note that the install_fips target wouldn't be useful just for developers. You would use that target to install the actual validated version of OpenSSL fips provider.

For that you would do first with the validated version:

```
./config ... enable-fips && make && make test && make install_fips
```

And then with updated bugfixed OpenSSL:
```
./config ... no-fips && make && make test && make install
```

Of course you could do `make install` in the first step instead but it would be kind of moot to have the installed library and other stuff immediately overwritten after that in the second step.

23. 4. 2021 19:02, 19:02, "Matthias St. Pierre" ***@***.***> napsal/a:
>
>> In other words if I spend the time to build the fips provider I most
>probably also want it installed if I am installing the openssl build.
>
***@***.*** I never thought of it that way and I must say I like your point of
>view very much: it makes perfect sense that the sequence
>
>```
>./config <options> <target> && make && make install
>```
>
>builds and installs openssl exactly with the features you have
>selected. Maybe I was just thinking too complicated. OTOH, for
>developers it might still be useful to have a low-level `install_fips`
>target similar to `install_sw` and `install_docs`. In view of what you
>said, I'm going to reconcider whether the change
>85ad7f9bc06f65e3f017cdaf2afdf876b56597ca is useful for developers or
>not.
>
>
>> And _maybe_ we should make the `fips` Configure option disabled by
>default.
>
>I also like your proposal to have the fips provider disabed by default.
>After all, enabling it has a heavy impact on build and install times
>and the majority of the countries in the world does not care so much
>about fips. We should talk about that on our next meeting.
>
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/openssl/openssl/pull/13684#issuecomment-825792393

rsbeckerca commented:
> 
> Please note that the install_fips target wouldn't be useful just for developers. You would use that target to install the actual validated version of OpenSSL fips provider. For that you would do first with the validated version: ``` ./config ... enable-fips && make && make test && make install_fips ``` And then with updated bugfixed OpenSSL: ``` ./config ... no-fips && make && make test && make install ``` Of course you could do `make install` in the first step instead but it would be kind of moot to have the installed library and other stuff immediately overwritten after that in the second step. 23. 4. 2021 19:02, 19:02, "Matthias St. Pierre" ***@***.***> napsal/a:
> > In other words if I spend the time to build the fips provider I most probably also want it installed if I am installing the openssl build.
> ***@***.*** I never thought of it that way and I must say I like your point of
> […](#)
> view very much: it makes perfect sense that the sequence ``` ./config <options> <target> && make && make install ``` builds and installs openssl exactly with the features you have selected. Maybe I was just thinking too complicated. OTOH, for developers it might still be useful to have a low-level `install_fips` target similar to `install_sw` and `install_docs`. In view of what you said, I'm going to reconcider whether the change [85ad7f9](https://github.com/openssl/openssl/commit/85ad7f9bc06f65e3f017cdaf2afdf876b56597ca) is useful for developers or not. > And _maybe_ we should make the `fips` Configure option disabled by default. I also like your proposal to have the fips provider disabed by default. After all, enabling it has a heavy impact on build and install times and the majority of the countries in the world does not care so much about fips. We should talk about that on our next meeting. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#13684 (comment)](https://github.com/openssl/openssl/pull/13684#issuecomment-825792393)

Our use case is somewhat different. I know it's not the best way, but there are some long-winded historical and political reasons for it. My team is building a non-validated FIPS distribution for our user community (a.k.a. NonStop x86). It is up to the end-customer or another vendor who wants to provide their own distribution with FIPS to validate their build. My team is specific about not doing the validation. When a customer takes our pre-built packages, they are on their own as far as whether to include FIPS or not. If they muck with FIPS, it obviously will not pass its load (correct?). What we may do, in light of this discussion, is provide a separately installable FIPS package.

Our use case will change - my sincere hope it does anyway - and have customers do their own builds to pick up fixes. This is the advice I have been pushing pretty hard for about a year with the community in social media articles. They will then be able to follow the above.

My ask is that this be well documented when it goes forward. Perhaps in a `NOTES-FIPS.md`. I'm happy to participate in that. The `NOTES-NONSTOP.md` file will get updated accordingly - I know who will do that.

mspncp commented:
> For that you would do first with the validated version:
> ```
> ./config ... enable-fips && make && make test && make install_fips
> ```

@t8m this scenario is the perfect argument for keeping commit 85ad7f9bc06f65e3f017cdaf2afdf876b56597ca, because it facilitates installing the fips provider without installing the entire software by [removing the dependency `install_fips: install_sw`](https://github.com/openssl/openssl/commit/85ad7f9bc06f65e3f017cdaf2afdf876b56597ca#diff-adf3ecfda4c92bad70382334b4d8089beb88119774ee3447c7a718df0de29ceaL474-R481).

mspncp commented:
@levitte currently the user gets a weird error message if he configured `no-fips` and calls `make install_fips` afterwards:

```
~/src/openssl$ make install_fips
make depend && make _build_sw
make[1]: Entering directory '/home/msp/src/openssl'
make[1]: Leaving directory '/home/msp/src/openssl'
make[1]: Entering directory '/home/msp/src/openssl'
make[1]: Leaving directory '/home/msp/src/openssl'
*** Installing FIPS module
install  -> /opt/openssl-dev/lib/ossl-modules/
cp: cannot stat '': No such file or directory
make: *** [Makefile:1554: install_fips] Error 1
```

I would like to conditionally replace the make targets `install_fips` and `uninstall_fips` 

https://github.com/openssl/openssl/blob/60caee4bf1c780e95e2853b785e94ef65c8f2621/Configurations/unix-Makefile.tmpl#L599-L616

by rules wich ouput a proper error message if `no-fips` was configured:

```
install_fips:
	@$(ECHO) "The 'install_fips' target requires the 'enable-fips' option"

uninstall_fips:
	@$(ECHO) "The 'uninstall_fips' target requires the 'enable-fips' option"
```

Could you please throw me a Perl code snippet how I could achieve this in the Makefile template?

rsbeckerca commented:
> Could you please throw me a Perl code snippet how I could achieve this in the Makefile template?

Would it make sense just to omit the two targets `install-fips` and `uninstall-fips`? Although the counter-argument is that this could be a bit cruel to the end user.

mspncp commented:
Ready for final review. Putting an @openssl/otc hold on this PR, because it changes the FIPS installation procedure significantly, so it deserves a discussion on our next weekly.

For more details, see the new documentation in commit https://github.com/openssl/openssl/pull/13684/commits/6b68e77cb5f7f78bd4f36e412735aaa331b02b96:

    README-FIPS: document the installation of the FIPS provider
    
    Note that configuration and installation procedure has changed:
    
    - The FIPS provider is now disabled by default and needs to
      be enabled by configuring with `enalbe-fips`.
    - If the FIPS provider is enabled, it gets installed automatically.
      There is no extra installation step required anymore.
    
    This is more natural and coincides with the expectation of the
    user, namely "what's configured, gets installed".

t8m reviewed (approved):
Good work!

t8m commented:
Heh, the approval was a little bit premature as the change of no-fips by default broke the 90-test_fipsload.t. Please fix it. However please also add enable-fips to most of the test runs maybe except one of those that failed currently - for example the basic_clang.

mspncp commented:
> Good work!

Thanks! It wouldn't have been nearly as good without your suggestions. :-)

t8m commented:
Probably also the external-tests and non-caching builds could go without enable-fips as those take more time and aren't that much relevant for the fips module.

mspncp commented:
> Heh, the approval was a little bit premature as the change of no-fips by default broke the 90-test_fipsload.t. Please fix it. However please also add enable-fips to most of the test runs maybe except one of those that failed currently - for example the basic_clang.

You are right, the general CI setup needs to be changed to add `enable-fips`. I'm not so familiar with the current GitHub Actions configuration after migrating from Travis. Maybe it would be good to have a separate Pull Request for the CI which goes in first? As for ` 90-test_fipsload.t`, I'll take care of it.

Anyway, I didn't intend to merge this PR before our weekly meeting tomorrow.

mspncp commented:
> You are right, the general CI setup needs to be changed to explicitly add `enable-fips`

Explicitly adding  `enable-fips` or ` no-fips` would be a good idea imho. Not only because it makes the CI less susceptible to a change of the default, it's also more expressive.

t8m commented:
See #15027 for the explicit enable-fips/no-fips.

mspncp commented:
The last force-push was a plain rebase without changes to the PR to resolve the conflicts in CHANGES.md. 

Since I need to rebase on master to resolve conflicts, I decided to cherry-pick your fixes from #15027 before rebasing.

mspncp commented:
The PR was discussed in the weekly meeting and no objections raised.

mspncp commented:
@t8m should I drop the CHANGES entry (commit 2f6ee30f56153e147b74ad90033594712c42f811))?

t8m commented:
> @t8m should I drop the CHANGES entry (commit [2f6ee30](https://github.com/openssl/openssl/commit/2f6ee30f56153e147b74ad90033594712c42f811))?

I think it should be modified to mention the appearance of the fips option in 3.0. I.e., adjust it so it describes the change from 1.1.1 branch.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

mspncp commented:
@openssl/otc te entire FIPS provider has now a single CHANGES entry, please take a look at commit dad7132e11db894c0bcd4feb11bb06b69e4e50da. I refrained from quoting each single team member explicitly and attributed it to **OpenSSL Team Members** instead. Is that ok for you?


> * OpenSSL includes a cryptographic module that is intended to be FIPS 140-2
>   validated. The module is implemented as an OpenSSL provider, the so-called
>   FIPS provider. A list of all changes related to the FIPS provider would go
>   beyond the scope of this CHANGES file, please consult the README-FIPS and
>   README-PROVIDERS files, as well as the migration guide.
>
>   The FIPS provider is disabled by default and needs to be enabled explicitly
>   at configuration time using the `enable-fips` option. If it is enabled,
>   the FIPS provider gets built and installed in addition to the default and
>   the legacy provider. No separate installation procedure is necessary.
>   There is however a dedicated `install_fips` make target, which serves the
>   special purpose of installing only the FIPS provider into an existing
>   OpenSSL installation.
>
>   *OpenSSL Team Members*

richsalz commented:
Prefer to see it as "OpenSSL Team and Contributors" since much important work was done by others, e.g., all of @slontis's efforts, and Team could imply just OMC OTC Fellows.

levitte commented:
> e.g., all of @slontis's efforts, and Team could imply just OMC OTC Fellows.

FYI, @slontis is an OTC member.  :wink:

mspncp commented:
Ok, but @richsalz 

> > e.g., all of @slontis's efforts, and Team could imply just OMC OTC Fellows.
> 
> FYI, @slontis is an OTC member. 😉

Ok, but @richsalz isn't a team member (anymore) and certainly deserves to be mentioned. I'll change it to 

> *OpenSSL team members and many third party contributors*

how does that read?

richsalz commented:
I like that change.  You could also say "The OpenSSL Authors" which shows up in the copyright notice.

As for @slontis, that'll teach me to try for humility :)

mspncp commented:
The term "OpenSSL Authors" too abstract and sounds too much like legal stuff imho. I'll stick to team members.

mspncp commented:
See commit 6691b3367fae250f5c4430349b7477f7d290fe94.

paulidale commented:
@mspncp I'm having some doubts in hindsight.

For people who are building OpenSSL exactly once and then shipping it, I don't think the build time is all that relevant.

t8m commented:
#15027 is now merged

mspncp commented:
> @mspncp I'm having some doubts in hindsight.

About what? About disabling 'fips' by default? Do you want me to hold that change back?

> For people who are building OpenSSL exactly once and then shipping it, I don't think the build time is all that relevant.

People don't care about build times does is not actually an argument. What is causing your doubts?

richsalz commented:
I think a concern is that someone edits a header file and then the right things don't recompile.  I think it's better to spend a few moments and make sure future builds are right.

rsbeckerca commented:
> 
> 
> > @mspncp I'm having some doubts in hindsight.
> 
> About what? About disabling 'fips' by default? Do you want me to hold that change back?
> 
> > For people who are building OpenSSL exactly once and then shipping it, I don't think the build time is all that relevant.
> 
> "People don't care about build times" is not actually an argument. (I'd rather say they do care if it's more than 200%) What is causing your doubts?

We run a complete clean prior to every definitive build, so `makedepend` for us is not useful. I would expect that would be a standard process for everyone - considering there have been issues of parts of OpenSSL not being rebuilt as expected without doing a clean.

mspncp commented:
> I think a concern is that someone edits a header file and then the right things don't recompile. I think it's better to spend a few moments and make sure future builds are right.

But the 'makedepend' change is in #15050.

mspncp commented:
> But the 'makedepend' change is in #15050.

Please discuss your concerns related to 'makedepend' there.

mspncp commented:
> #15027 is now merged

I dropped your two comit and rebased to the head of the master branch. There were no conflicts, no review is required and the grace period is not reset imho.

mspncp commented:
Nevertheless, I'll wait for feedback from @paulidale before merging.

rsbeckerca commented:
Don't know whether it helps, but commit 4189dc3782 looks good on NonStop x86 - no other place to report that yet.

t8m commented:
I think this was discussed enough on the OTC meeting and I do not see any serious concerns against this PR being raised here. Doing makedepend by default or not is a different issue and PR. This should be merged.

t8m commented:
@mspncp I got confirmation from @paulidale that his comment was about disabling the makedepend. So this is OK to be merged.

mspncp commented:
I was busy elsewhere, but I'm going to merge it now.

mspncp commented:
Merged to master. Thank you for reviewing and your valuable suggestions @t8m!

* 3b9e47695f CHANGES: document the FIPS provider configuration and installation
* f2ea01d9f1 README-FIPS: document the installation of the FIPS provider
* b2d8c7b6a3 Configure: disable fips mode by default
* afa0a13c1a Configure: sort the disablables alphabetically
* d9ce268151 build.info: add the Perl wrapper to build generator programs on Windows
* 18da9fc31f Configure/Makefile: install the fips provider if it was configured
* 4e282708c5 Configure/Makefile: don't generate a fresh fipsmodule.cnf when installing it
* 5b68918185 Configure/Makefile: separate install of the FIPS module
* c3bda8a2e0 Configure/Makefile: correct the FIPS module configuration file path
* b6821df0d0 Configure/Makefile: use the correct openssl app for FIPS installation
* 59cf286919 Configure/Makefile: fix the `-macopt` argument of the fipsinstall command

csim-20 commented:
Block meta
